# Capstone Project - The Battle of the Neighborhoods (Week 2)

### Applied Data Science Capstone by IBM/Coursera

## Introduction & Business Problem

The purpose of this project is to determine the best location to open up a high-end cafe in Houston. The definition of a high-end cafe is one that caters to the upper-middle to upper financial class of society. Goods and services will have a premium that is justified by the high quality in comparison to general cafes. Products from the cafe such as baked goods and beverages will be created by distinguished bakers and dessert chefs. The ingredients used will have a greater focus on quality than on affordability. As the economy thrives and a larger number of people realize a personal taste for premium services and goods, intelligent investors will be looking for locales that cater to a higher echelon of society (financially). It is innate human nature that people with greater than average buying power are willing to pay a greater premium for exclusivity. The Foursquare location data will be vital to determining locations that are dense, wealthy residential neighborhoods that lack establishments of a similar nature. The greater the competition, the less appealing a certain location will be. 

## Data

As mentioned earlier, the Foursquare data will be essential to determine the number of cafes (especially ones that tailor to the same market as mine will) in all neighborhoods in Houston. This will enable me to make an educated decision on which neighborhoods to filter out and which ones are prime candidates based on market availability. Another source I will use is median household income data per zipcode in Houston. This table also contains population per zipcode as well as specific latitude and longitude information (location). Knowing this, I can determine which neighborhoods have the most residents with the ability to afford the products and services my establishment will be offering, and which neighborhoods I will have to filter out due to a lower level of buying power. Coupling this location/financial data with the Foursquare data I will be able to cluster specific zipcodes/neighborhoods within Houston that are the most ideal candidates for beginning a upscale cafe business. The specific URL of this dataset is the following: "http://zipatlas.com/us/tx/houston/zip-code-comparison/median-household-income.htm". An example of the data is the following: zipcode 77010 of Houston, Texas has a population of 76 people with an average household income of $200,000. The location of this neighborhood is 29.75310, -95.361109.

Based on definition of our problem, factors that will influence our decision are:

number of existing cafes in the neighborhood (high-end cafes)

Following data sources will be needed to extract/generate the required information:

number of restaurants and their type and location in every neighborhood will be obtained using Foursquare API

### Exploring Houston's Neighborhoods

Let's start off by importing all of the necessary tools to solve this problem.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from urllib.request import urlopen
from bs4 import BeautifulSoup

Now let's open the website (source) and begin extraction of data that we want. A simple test is important to verify that our code is working and we are looking at the correct dataset.

In [42]:
data = pd.read_html('http://zipatlas.com/us/tx/houston/zip-code-comparison/median-household-income.htm')
isolate_data=data[-3]
isolate_data

0         1                      2               3           4  \
0     #  Zip Code               Location            City  Population   
1    1.     77010  29.754310, -95.361109  Houston, Texas          76   
2    2.     77094  29.769285, -95.681292  Houston, Texas        7779   
3    3.     77046  29.733084, -95.430659  Houston, Texas         471   
4    4.     77059  29.615219, -95.134960  Houston, Texas       16690   
5    5.     77005  29.718435, -95.423555  Houston, Texas       23338   
6    6.     77024  29.771991, -95.515453  Houston, Texas       32746   
7    7.     77068  30.008830, -95.487234  Houston, Texas        9505   
8    8.     77095  29.916055, -95.663077  Houston, Texas       39275   
9    9.     77062  29.575781, -95.134334  Houston, Texas       26978   
10  10.     77056  29.749035, -95.469021  Houston, Texas       14031   
11  11.     77079  29.774523, -95.601071  Houston, Texas       30887   
12  12.     77069  29.986925, -95.524903  Houston, Texas       14756   
13  13.     77019  29.752702, -95.407379  Houston, Texas       15640   
14  14.     77064  29.924216, -95.537411  Houston, Texas       35966   
15  15.     77070  29.979670, -95.572835  Houston, Texas       32385   
16  16.     77027  29.744002, -95.443213  Houston, Texas       14217   
17  17.     77065  29.926556, -95.603242  Houston, Texas       27070   
18  18.     77041  29.857989, -95.585771  Houston, Texas       25275   
19  19.     77066  29.958498, -95.500223  Houston, Texas       27638   
20  20.     77030  29.704849, -95.404429  Houston, Texas       10984   
21  21.     77089  29.588469, -95.218012  Houston, Texas       36450   
22  22.     77084  29.826860, -95.650860  Houston, Texas       63557   
23  23.     77083  29.693991, -95.647830  Houston, Texas       52061   
24  24.     77077  29.749158, -95.620344  Houston, Texas       42416   
25  25.     77040  29.876755, -95.532143  Houston, Texas       39520   
26  26.     77058  29.561579, -95.099626  Houston, Texas       15206   
27  27.     77025  29.688811, -95.434780  Houston, Texas       23581   
28  28.     77098  29.734813, -95.416098  Houston, Texas       12179   
29  29.     77057  29.744068, -95.489210  Houston, Texas       35491   
..  ...       ...                    ...             ...         ...   
67  67.     77039  29.911171, -95.341182  Houston, Texas       27718   
68  68.     77074  29.687947, -95.515725  Houston, Texas       39159   
69  69.     77092  29.830858, -95.474872  Houston, Texas       37899   
70  70.     77013  29.791808, -95.228991  Houston, Texas       18378   
71  71.     77017  29.692600, -95.252270  Houston, Texas       32823   
72  72.     77061  29.647012, -95.276630  Houston, Texas       25500   
73  73.     77087  29.685621, -95.304159  Houston, Texas       36194   
74  74.     77029  29.764519, -95.265896  Houston, Texas       18272   
75  75.     77023  29.721825, -95.318275  Houston, Texas       33050   
76  76.     77076  29.860021, -95.382184  Houston, Texas       30030   
77  77.     77060  29.935512, -95.397624  Houston, Texas       36306   
78  78.     77078  29.845618, -95.239696  Houston, Texas       14479   
79  79.     77048  29.618714, -95.324221  Houston, Texas       14267   
80  80.     77036  29.699048, -95.536507  Houston, Texas       76146   
81  81.     77033  29.666880, -95.335133  Houston, Texas       27676   
82  82.     77093  29.863400, -95.341409  Houston, Texas       45908   
83  83.     77081  29.712099, -95.480935  Houston, Texas       49691   
84  84.     77032  29.987805, -95.353412  Houston, Texas       10796   
85  85.     77091  29.853067, -95.435966  Houston, Texas       24460   
86  86.     77011  29.742988, -95.307114  Houston, Texas       23052   
87  87.     77012  29.717567, -95.275115  Houston, Texas       25174   
88  88.     77016  29.863166, -95.298522  Houston, Texas       29753   
89  89.     77022  29.831590, -95.379553  Houston, Texas       31781   
90  90.     77021  29.697363, -95.358898  Hous

Let's rename the headers with the data in the first row of the data, and remove the first column since it's redundant with the index. Reorganize table to most efficient and comfortable form for further use.

In [43]:
isolate_data=isolate_data.rename(columns=isolate_data.iloc[0]).drop(isolate_data.index[0])

In [44]:
dfwant=isolate_data[['Zip Code','Location','Population','Avg. Income/H/hold']].copy()
dfwant

Zip Code               Location Population Avg. Income/H/hold
1     77010  29.754310, -95.361109         76        $200,000.00
2     77094  29.769285, -95.681292       7779        $123,244.00
3     77046  29.733084, -95.430659        471        $105,863.00
4     77059  29.615219, -95.134960      16690        $104,844.00
5     77005  29.718435, -95.423555      23338        $104,035.00
6     77024  29.771991, -95.515453      32746         $82,620.00
7     77068  30.008830, -95.487234       9505         $77,724.00
8     77095  29.916055, -95.663077      39275         $76,814.00
9     77062  29.575781, -95.134334      26978         $75,689.00
10    77056  29.749035, -95.469021      14031         $71,926.00
11    77079  29.774523, -95.601071      30887         $70,587.00
12    77069  29.986925, -95.524903      14756         $68,662.00
13    77019  29.752702, -95.407379      15640         $61,899.00
14    77064  29.924216, -95.537411      35966         $60,945.00
15    77070  29.979670, -95.572835      32385         $60,887.00
16    77027  29.744002, -95.443213      14217         $60,032.00
17    77065  29.926556, -95.603242      27070         $58,426.00
18    77041  29.857989, -95.585771      25275         $56,746.00
19    77066  29.958498, -95.500223      27638         $55,760.00
20    77030  29.704849, -95.404429      10984         $53,199.00
21    77089  29.588469, -95.218012      36450         $53,123.00
22    77084  29.826860, -95.650860      63557         $53,075.00
23    77083  29.693991, -95.647830      52061         $52,931.00
24    77077  29.749158, -95.620344      42416         $52,667.00
25    77040  29.876755, -95.532143      39520         $50,183.00
26    77058  29.561579, -95.099626      15206         $48,744.00
27    77025  29.688811, -95.434780      23581         $48,232.00
28    77098  29.734813, -95.416098      12179         $47,703.00
29    77057  29.744068, -95.489210      35491         $46,563.00
30    77096  29.674133, -95.478354      33987         $46,140.00
..      ...                    ...        ...                ...
67    77039  29.911171, -95.341182      27718         $32,310.00
68    77074  29.687947, -95.515725      39159         $32,238.00
69    77092  29.830858, -95.474872      37899         $32,095.00
70    77013  29.791808, -95.228991      18378         $31,834.00
71    77017  29.692600, -95.252270      32823         $31,762.00
72    77061  29.647012, -95.276630      25500         $30,440.00
73    77087  29.685621, -95.304159      36194         $30,346.00
74    77029  29.764519, -95.265896      18272         $30,308.00
75    77023  29.721825, -95.318275      33050         $29,698.00
76    77076  29.860021, -95.382184      30030         $28,270.00
77    77060  29.935512, -95.397624      36306         $27,867.00
78    77078  29.845618, -95.239696      14479         $27,556.00
79    77048  29.618714, -95.324221      14267         $27,391.00
80    77036  29.699048, -95.536507      76146         $26,931.00
81    77033  29.666880, -95.335133      27676         $26,544.00
82    77093  29.863400, -95.341409      45908         $26,270.00
83    77081  29.712099, -95.480935      49691         $26,026.00
84    77032  29.987805, -95.353412      10796         $25,713.00
85    77091  29.853067, -95.435966      24460         $25,405.00
86    77011  29.742988, -95.307114      23052         $24,612.00
87    77012  29.717567, -95.275115      25174         $23,910.00
88    77016  29.863166, -95.298522      29753         $23,835.00
89    77022  29.831590, -95.379553      31781         $23,746.00
90    77021  29.697363, -95.358898      24159         $21,950.00
91    77020  29.773179, -95.314327      28679         $21,864.00
92    77028  29.825122, -95.286360      16490         $21,131.00
93    77004  29.724893, -95.363752      30379         $20,840.00
94    77003  29.749563, -95.346265       9195         $19,252.00
95    77051  29.663567, -95.375573      13235         $17,529.00
96    77026  29.800187, -95.328888    

In [45]:
dfwant['Location']=dfwant['Location'].str.strip()

In [46]:
dfwant[['Latitude','Longitude']] = dfwant.Location.str.split(",",expand=True,)

In [47]:
dfwant=dfwant.drop(columns=['Location'])
dfwant

Zip Code Population Avg. Income/H/hold   Latitude    Longitude
1     77010         76        $200,000.00  29.754310   -95.361109
2     77094       7779        $123,244.00  29.769285   -95.681292
3     77046        471        $105,863.00  29.733084   -95.430659
4     77059      16690        $104,844.00  29.615219   -95.134960
5     77005      23338        $104,035.00  29.718435   -95.423555
6     77024      32746         $82,620.00  29.771991   -95.515453
7     77068       9505         $77,724.00  30.008830   -95.487234
8     77095      39275         $76,814.00  29.916055   -95.663077
9     77062      26978         $75,689.00  29.575781   -95.134334
10    77056      14031         $71,926.00  29.749035   -95.469021
11    77079      30887         $70,587.00  29.774523   -95.601071
12    77069      14756         $68,662.00  29.986925   -95.524903
13    77019      15640         $61,899.00  29.752702   -95.407379
14    77064      35966         $60,945.00  29.924216   -95.537411
15    77070      32385         $60,887.00  29.979670   -95.572835
16    77027      14217         $60,032.00  29.744002   -95.443213
17    77065      27070         $58,426.00  29.926556   -95.603242
18    77041      25275         $56,746.00  29.857989   -95.585771
19    77066      27638         $55,760.00  29.958498   -95.500223
20    77030      10984         $53,199.00  29.704849   -95.404429
21    77089      36450         $53,123.00  29.588469   -95.218012
22    77084      63557         $53,075.00  29.826860   -95.650860
23    77083      52061         $52,931.00  29.693991   -95.647830
24    77077      42416         $52,667.00  29.749158   -95.620344
25    77040      39520         $50,183.00  29.876755   -95.532143
26    77058      15206         $48,744.00  29.561579   -95.099626
27    77025      23581         $48,232.00  29.688811   -95.434780
28    77098      12179         $47,703.00  29.734813   -95.416098
29    77057      35491         $46,563.00  29.744068   -95.489210
30    77096      33987         $46,140.00  29.674133   -95.478354
..      ...        ...                ...        ...          ...
67    77039      27718         $32,310.00  29.911171   -95.341182
68    77074      39159         $32,238.00  29.687947   -95.515725
69    77092      37899         $32,095.00  29.830858   -95.474872
70    77013      18378         $31,834.00  29.791808   -95.228991
71    77017      32823         $31,762.00  29.692600   -95.252270
72    77061      25500         $30,440.00  29.647012   -95.276630
73    77087      36194         $30,346.00  29.685621   -95.304159
74    77029      18272         $30,308.00  29.764519   -95.265896
75    77023      33050         $29,698.00  29.721825   -95.318275
76    77076      30030         $28,270.00  29.860021   -95.382184
77    77060      36306         $27,867.00  29.935512   -95.397624
78    77078      14479         $27,556.00  29.845618   -95.239696
79    77048      14267         $27,391.00  29.618714   -95.324221
80    77036      76146         $26,931.00  29.699048   -95.536507
81    77033      27676         $26,544.00  29.666880   -95.335133
82    77093      45908         $26,270.00  29.863400   -95.341409
83    77081      49691         $26,026.00  29.712099   -95.480935
84    77032      10796         $25,713.00  29.987805   -95.353412
85    77091      24460         $25,405.00  29.853067   -95.435966
86    77011      23052         $24,612.00  29.742988   -95.307114
87    77012      25174         $23,910.00  29.717567   -95.275115
88    77016      29753         $23,835.00  29.863166   -95.298522
89    77022      31781         $23,746.00  29.831590   -95.379553
90    77021      24159         $21,950.00  29.697363   -95.358898
91    77020      28679         $21,864.00  29.773179   -95.314327
92    77028      16490         $21,131.00  29.825122   -95.286360
93    77004      30379         $20,840.00  29.724893   -95.363752
94    77003       9195         $19,252.00  29.749563   -95.346265
95    77051      13235         $17,529.00  

In [48]:
dfwant['Avg. Income/H/hold'] = dfwant['Avg. Income/H/hold'].str.replace(',', '')
dfwant['Avg. Income/H/hold'] = dfwant['Avg. Income/H/hold'].str.replace('$', '')
dfwant['Avg. Income/H/hold'] = dfwant['Avg. Income/H/hold'].str[:-3]
dfwant['Avg. Income/H/hold'] = dfwant['Avg. Income/H/hold'].astype(int)
dfwant['Zip Code'] = dfwant['Zip Code'].astype(int)
dfwant['Population'] = dfwant['Population'].astype(int)
dfwant['Latitude'] = dfwant['Latitude'].astype(float)
dfwant['Longitude'] = dfwant['Longitude'].astype(float)

In [49]:
dfwant

Zip Code  Population  Avg. Income/H/hold   Latitude  Longitude
1      77010          76              200000  29.754310 -95.361109
2      77094        7779              123244  29.769285 -95.681292
3      77046         471              105863  29.733084 -95.430659
4      77059       16690              104844  29.615219 -95.134960
5      77005       23338              104035  29.718435 -95.423555
6      77024       32746               82620  29.771991 -95.515453
7      77068        9505               77724  30.008830 -95.487234
8      77095       39275               76814  29.916055 -95.663077
9      77062       26978               75689  29.575781 -95.134334
10     77056       14031               71926  29.749035 -95.469021
11     77079       30887               70587  29.774523 -95.601071
12     77069       14756               68662  29.986925 -95.524903
13     77019       15640               61899  29.752702 -95.407379
14     77064       35966               60945  29.924216 -95.537411
15     77070       32385               60887  29.979670 -95.572835
16     77027       14217               60032  29.744002 -95.443213
17     77065       27070               58426  29.926556 -95.603242
18     77041       25275               56746  29.857989 -95.585771
19     77066       27638               55760  29.958498 -95.500223
20     77030       10984               53199  29.704849 -95.404429
21     77089       36450               53123  29.588469 -95.218012
22     77084       63557               53075  29.826860 -95.650860
23     77083       52061               52931  29.693991 -95.647830
24     77077       42416               52667  29.749158 -95.620344
25     77040       39520               50183  29.876755 -95.532143
26     77058       15206               48744  29.561579 -95.099626
27     77025       23581               48232  29.688811 -95.434780
28     77098       12179               47703  29.734813 -95.416098
29     77057       35491               46563  29.744068 -95.489210
30     77096       33987               46140  29.674133 -95.478354
..       ...         ...                 ...        ...        ...
67     77039       27718               32310  29.911171 -95.341182
68     77074       39159               32238  29.687947 -95.515725
69     77092       37899               32095  29.830858 -95.474872
70     77013       18378               31834  29.791808 -95.228991
71     77017       32823               31762  29.692600 -95.252270
72     77061       25500               30440  29.647012 -95.276630
73     77087       36194               30346  29.685621 -95.304159
74     77029       18272               30308  29.764519 -95.265896
75     77023       33050               29698  29.721825 -95.318275
76     77076       30030               28270  29.860021 -95.382184
77     77060       36306               27867  29.935512 -95.397624
78     77078       14479               27556  29.845618 -95.239696
79     77048       14267               27391  29.618714 -95.324221
80     77036       76146               26931  29.699048 -95.536507
81     77033       27676               26544  29.666880 -95.335133
82     77093       45908               26270  29.863400 -95.341409
83     77081       49691               26026  29.712099 -95.480935
84     77032       10796               25713  29.987805 -95.353412
85     77091       24460               25405  29.853067 -95.435966
86     77011       23052               24612  29.742988 -95.307114
87     77012       25174               23910  29.717567 -95.275115
88     77016       29753               23835  29.863166 -95.298522
89     77022       31781               23746  29.831590 -95.379553
90     77021       24159               21950  29.697363 -95.358898
91     77020       28679               21864  29.773179 -95.314327
92     77028       16490               21131  29.825122 -95.286360
93     77004       30379               20840  29.724893 -95.363752
94     77003        9195               19252  29.74

Isolate the neighborhoods with the Top 5 highest average income per household since this is the market I want to be targeting.

In [51]:
dfrich = dfwant[dfwant['Avg. Income/H/hold'] > 100000]
dfrich

Zip Code  Population  Avg. Income/H/hold   Latitude  Longitude
1     77010          76              200000  29.754310 -95.361109
2     77094        7779              123244  29.769285 -95.681292
3     77046         471              105863  29.733084 -95.430659
4     77059       16690              104844  29.615219 -95.134960
5     77005       23338              104035  29.718435 -95.423555

### Foursquare

Shift to Foursquare to visualize the Houston area.

In [12]:
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0           conda-forge
    geopy:          

In [13]:
address = 'Houston, TX'

geolocator = Nominatim(user_agent="Houston_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Houston are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Houston are 29.7589382, -95.3676974.


In [14]:
# create map of Houston using latitude and longitude values
map_houston = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(dfrich['Latitude'], dfrich['Longitude'], dfrich['Zip Code']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_houston)  
    
map_houston

## Methodology

Create a function to summon all nearby venues around the select zip codes of Houston.

In [15]:
import requests

In [16]:
def getNearbyVenues(names, latitudes, longitudes, radius=1600):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Zip Code', 
                  'Zip Code Latitude', 
                  'Zip Code Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Input Foursquare credentials to utilize API

In [69]:
CLIENT_ID='OGAJTDQP2NS0OBOGZDU4BENV15GWD5RGEYI4C0TN4ETTB2OP'
CLIENT_SECRET='LRVL3KUWY4EX2GZZYIQADLESIGCVPJG3EBXGAJ1CYTIR00VJ'
VERSION='20180605'
LIMIT=100
radius=1000

In [72]:
houston_venues=getNearbyVenues(names=dfrich['Zip Code'],
                              latitudes=dfrich['Latitude'],
                              longitudes=dfrich['Longitude']
                              )

77010
77094
77046
77059
77005


In [73]:
houston_venues.groupby(['Zip Code']).count()

Zip Code Latitude  Zip Code Longitude  Venue  Venue Latitude  \
Zip Code                                                                 
77005                   100                 100    100             100   
77010                   100                 100    100             100   
77046                   100                 100    100             100   
77059                    17                  17     17              17   
77094                     4                   4      4               4   

          Venue Longitude  Venue Category  
Zip Code                                   
77005                 100             100  
77010                 100             100  
77046                 100             100  
77059                  17              17  
77094                   4               4

In [74]:
print('There are {} uniques categories.'.format(len(houston_venues['Venue Category'].unique())))

There are 121 uniques categories.


Let's take a careful look at the amenities that are located within each, unique zip code.

In [75]:
#one hot encoding
houston_onehot=pd.get_dummies(houston_venues[['Venue Category']], prefix="",prefix_sep="")

# add neighborhood column back to dataframe
houston_onehot['Zip Code'] = houston_venues['Zip Code'] 

# move neighborhood column to the first column
fixed_columns = [houston_onehot.columns[-1]] + list(houston_onehot.columns[:-1])
houston_onehot = houston_onehot[fixed_columns]

houston_onehot.head()

Zip Code  ATM  American Restaurant  Arts & Crafts Store  Asian Restaurant  \
0     77010    0                    0                    0                 0   
1     77010    0                    0                    0                 0   
2     77010    0                    0                    0                 0   
3     77010    0                    0                    0                 0   
4     77010    0                    0                    0                 0   

   Auto Dealership  Auto Garage  BBQ Joint  Bagel Shop  Bakery  ...  \
0                0            0          0           0       0  ...   
1                0            0          0           0       0  ...   
2                0            0          0           0       0  ...   
3                0            0          0           0       0  ...   
4                0            0          0           0       0  ...   

   Thai Restaurant  Theater  Trail  Turkish Restaurant  Vietnamese Restaurant  \
0                0        0      0                   0                      0   
1                0        0      0                   0                      0   
2                0        0      0                   0                      0   
3                0        0      0                   0                      0   
4                0        0      0                   0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Yoga Studio  
0                0         0          0            0            0  
1                0         0          0            0            0  
2                0         0          0            0            0  
3                0         0          0            0            0  
4                0         0          0            0            0  

[5 rows x 122 columns]

In [76]:
houston_grouped = houston_onehot.groupby('Zip Code').mean().reset_index()
houston_grouped['Zip Code'] = houston_grouped['Zip Code'].astype(str)
houston_grouped

Zip Code       ATM  American Restaurant  Arts & Crafts Store  \
0    77005  0.000000                 0.05                 0.00   
1    77010  0.000000                 0.03                 0.01   
2    77046  0.000000                 0.03                 0.00   
3    77059  0.058824                 0.00                 0.00   
4    77094  0.000000                 0.00                 0.00   

   Asian Restaurant  Auto Dealership  Auto Garage  BBQ Joint  Bagel Shop  \
0          0.000000             0.00     0.000000       0.01        0.01   
1          0.000000             0.00     0.000000       0.03        0.00   
2          0.020000             0.00     0.000000       0.01        0.00   
3          0.058824             0.00     0.058824       0.00        0.00   
4          0.000000             0.25     0.000000       0.00        0.00   

   Bakery  ...  Thai Restaurant  Theater  Trail  Turkish Restaurant  \
0    0.01  ...             0.00     0.01   0.01                0.01   
1    0.00  ...             0.01     0.02   0.00                0.00   
2    0.03  ...             0.01     0.00   0.00                0.00   
3    0.00  ...             0.00     0.00   0.00                0.00   
4    0.00  ...             0.00     0.00   0.00                0.00   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                   0.00             0.00      0.02       0.01     0.000000   
1                   0.04             0.00      0.02       0.00     0.000000   
2                   0.00             0.01      0.02       0.00     0.000000   
3                   0.00             0.00      0.00       0.00     0.058824   
4                   0.00             0.00      0.00       0.00     0.000000   

   Yoga Studio  
0         0.01  
1         0.00  
2         0.01  
3         0.00  
4         0.00  

[5 rows x 122 columns]

In [77]:
num_top_venues = 5

for hood in houston_grouped['Zip Code']:
    print("----"+hood+"----")
    temp = houston_grouped[houston_grouped['Zip Code'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----77005----
                 venue  freq
0          Coffee Shop  0.05
1  American Restaurant  0.05
2   Italian Restaurant  0.05
3         Burger Joint  0.04
4       Ice Cream Shop  0.03


----77010----
                   venue  freq
0                  Hotel  0.09
1            Pizza Place  0.04
2  Vietnamese Restaurant  0.04
3                    Bar  0.04
4                   Park  0.04


----77046----
                     venue  freq
0       Mexican Restaurant  0.07
1             Burger Joint  0.05
2       Italian Restaurant  0.05
3       Seafood Restaurant  0.05
4  New American Restaurant  0.04


----77059----
                venue  freq
0         Pizza Place  0.12
1                 ATM  0.06
2       Grocery Store  0.06
3  Salon / Barbershop  0.06
4      Sandwich Place  0.06


----77094----
                        venue  freq
0  Construction & Landscaping  0.25
1             Auto Dealership  0.25
2                 Gas Station  0.25
3           Other Repair Shop  0.25
4               

In [78]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now that we have all of the information we need, it is time to organize the data in a viewable format. Append the top 5 venues for each candidate location to the original dataframe to have a comprehensive view of the information I have collected.

In [85]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Zip Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Zip Code'] = houston_grouped['Zip Code']

for ind in np.arange(houston_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(houston_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Zip Code       1st Most Common Venue 2nd Most Common Venue  \
0    77005         American Restaurant    Italian Restaurant   
1    77010                       Hotel                   Bar   
2    77046          Mexican Restaurant    Seafood Restaurant   
3    77059                 Pizza Place                   Spa   
4    77094  Construction & Landscaping       Auto Dealership   

  3rd Most Common Venue  4th Most Common Venue    5th Most Common Venue  
0           Coffee Shop           Burger Joint           Ice Cream Shop  
1           Pizza Place  Vietnamese Restaurant                     Park  
2          Burger Joint     Italian Restaurant  New American Restaurant  
3          Home Service            Wings Joint       Italian Restaurant  
4     Other Repair Shop            Gas Station     Fast Food Restaurant

In [86]:
# set number of clusters
kclusters = 5

houston_grouped_clustering = houston_grouped.drop('Zip Code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(houston_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([3, 4, 1, 2, 0], dtype=int32)

Necessary datatype conversion to properly merge the two dataframes.

In [87]:
neighborhoods_venues_sorted["Zip Code"]=neighborhoods_venues_sorted["Zip Code"].astype(str).astype(int)

In [88]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

houston_merged = dfrich

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
houston_merged = houston_merged.join(neighborhoods_venues_sorted.set_index('Zip Code'), on='Zip Code')

houston_merged.head() # check the last columns!

Zip Code  Population  Avg. Income/H/hold   Latitude  Longitude  \
1     77010          76              200000  29.754310 -95.361109   
2     77094        7779              123244  29.769285 -95.681292   
3     77046         471              105863  29.733084 -95.430659   
4     77059       16690              104844  29.615219 -95.134960   
5     77005       23338              104035  29.718435 -95.423555   

   Cluster Labels       1st Most Common Venue 2nd Most Common Venue  \
1               4                       Hotel                   Bar   
2               0  Construction & Landscaping       Auto Dealership   
3               1          Mexican Restaurant    Seafood Restaurant   
4               2                 Pizza Place                   Spa   
5               3         American Restaurant    Italian Restaurant   

  3rd Most Common Venue  4th Most Common Venue    5th Most Common Venue  
1           Pizza Place  Vietnamese Restaurant                     Park  
2     Other Repair Shop            Gas Station     Fast Food Restaurant  
3          Burger Joint     Italian Restaurant  New American Restaurant  
4          Home Service            Wings Joint       Italian Restaurant  
5           Coffee Shop           Burger Joint           Ice Cream Shop

Finally, let's visualize these clusters on the map of Houston.

In [89]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(houston_merged['Latitude'], houston_merged['Longitude'], houston_merged['Zip Code'], houston_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Results & Discussion

The original datasource that was utilized in this project contained an extensive list of zip codes within the Houston area that were ranked by average household income. I targeted the top 5 zip codes based on a personal assumption that yearly incomes of over $100,000 would indicate a certain level of spending and taste for high-end services and products.

At first glance, this thought process would lead one to believe the area pertaining to the zip code 77010 is the most ideal location for an upscale cafe due to the fact that it has the highest average household income out of the entire list. However, it is important to take into consideration that for a retail operation, buying power is secondary to customer attendance. What this means is that while the ability for the surrounding population to buy the services and goods that I am offering is important, the frequency of customers is by far the most important factor. As such, to best garner a larger consumer base, 77010 is not an option.

Next, we have to take a careful look at the surrounding venues. Foursquare data accurately depicts that zipcodes 77094 and 77005 are not ideal locations for a coffee shop. 77005 simply because there are already a high density of cafes within this region. 77094 lacks the appeal due to the fact that the surrounding area is dominated by service providers that are uncharacteristic of a typical, high-income residential area. The atmosphere and vibe of a cafe is its lifeforce and being surrounded by gas stations, construction and landscaping, auto dealerships, and other repair shops would be difficult to justify for high-end consumers. That leaves two candidates - 77046 and 77059.

Taking a look at the surrounding venues in zipcode 77059, one can realize that there is a similar issue to 77094. While, admittedly, 77059 does have a significantly larger population of food providers in the vicinity, the price level and target population is more of the general public. The area is dominated by cheap and quick eats that cater to the average population. The people that I am targeting with this upscale cafe do not frequent these areas so the synergy with the surrounding venues would be nonexistent for my establishment.

The reason 77046 is the perfect location for an establishment of this nature is due to two factors. The first factor is the surrounding venues. According to the data drawn from Foursquare, the top 5 most common venues in this location are all food providers (restaurants) of varying nature - Italian, New American, Mexican and Seafood. This variety facilitates the perfect location for a cafe to come in. The synergy between an upscale dessert cafe in a neighborhood dominated by middle to upscale restaurants would create the perfect location for families to stop by and spend their time and money for leisurely or formal enjoyment. Consumers wouldn't have to travel farther than a kilometer to be faced with a variety of delectable options for any day and any occasion. The second factor is the location relative to local landmarks. This zipcode is located within the trifecta of Houston's most well-off areas - the galleria, river oaks, and Rice University. This is a strategically superior location to any other zipcode due to this placement as it could draw on the collective population of all three areas, which have plenty of consumers with plenty of buying power.

## Conclusion

The purpose of this project was to identify areas around Houston with a low number of cafes (particularly upscale establishments) in order to aid stakeholders in narrowing down the search for the optimal location to build a new high-end bakery/cafe. By sorting zip codes within Houston based on average household income, we have first narrowed down an extensive list of potential locations throughout Houston that justify further analysis. Then, we proceeded to access Foursquare data to identify what the characteristics of each of these locations were like. The information was primarily concerning what were the five most common venues in each of the zip codes. Clustering of each of these locations was thenm performed in order to visualize these points of interest on a map of Houston pending final decision making by stakeholders.

The final decision on an optimal cafe location will be made by stakeholders based on the unique characteristics of each zipcode location. The process will take into consideration additional factors such as the appeal of each region (proximity to high-end neighborhoods), levels of human traffic, as well as social and economic dynamics based on most common venues. The recommendation of this project is that the establishment be built in the location of zipcode 77046.